crypto/provider_predefined.h

```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core.h>
+
+static struct {
+    const char *name;
+    OSSL_provider_init_fn *init;
+    unsigned int is_fallback:1;
+} const predefined_providers[] = {
```

Comment:
Is it necessary to populate this array in an _header_ file?

How about having this in the header:
```
typedef struct {
...
} PREDEFINED_PROVIDER;

extern const PREDEFINED_PROVIDER predefined_providers[];
extern const size_t num_predefined_providers;
```

And this in a corresponding .c file:
```
const PREDEFINED_PROVIDER predefined_providers[] = {
    { "default", ossl_default_provider_init, 1 },
};

const size_t num_predefined_providers = OSSL_NELEM(predefined_providers);
```

Reply:
It's practical for generation purposes, and it's not like any other C unit should be interested in this particular file...

But hey, I could just put it all in `provider_core.c` if you prefer

Reply:
I guess it doesn't matter much which way.

I cringe a bit at code/data in headers, unless they're machine generated.

Reply:
> I cringe a bit at code/data in headers, *unless they're machine generated.*

And that's precisely it. We might not generate it now, but this prepares us for a future where we do.

Reply:
Okay, all fine if that's intended at some point.

Reply:
I agree that it jars putting data in a header file and requires nobody else to ever include it. I think it would avoid questions if you moved it.

Reply:
It's not like this is anything new...  see `crypto/asn1/charmap.h`, `crypto/asn1/asn1_item_list.h`, `crypto/asn1/standard_methods.h`, `crypto/asn1/tbl_standard.h`, `crypto/x509v3/standard_exts.h`

But yeahok, it can be made into a .c file